File: python/examples/langchain-krewe-inference-chatbot/chatbot.py
"""LangChain chatbot example: AgentKit + krewe x402-paywalled inference.

The agent has a CDP-managed Base wallet and four pay-per-call inference tools
through krewe (https://www.krewe.world), accessed via AgentKit's built-in
x402 action provider. Each call settles per-request in USDC on Base
(EIP-3009 ``transferWithAuthorization``) — no API keys, no provisioning.

Pricing (USDC base units / 6 decimals):
  text.structure  $0.005    text.embed       $0.01
  web.scrape      $0.02     text.complete    $0.05

Every USDC paid here flows back to the miners that answered the job via
krewe's hourly USDC → $KREW buyback recycler. Closes the circular economy.
"""

import os
import sys
import time

from coinbase_agentkit import (
    AgentKit,
    AgentKitConfig,
    CdpEvmWalletProvider,
    CdpEvmWalletProviderConfig,
    X402Config,
    cdp_api_action_provider,
    cdp_evm_wallet_action_provider,
    erc20_action_provider,
    wallet_action_provider,
    x402_action_provider,
)
from coinbase_agentkit_langchain import get_langchain_tools
from dotenv import load_dotenv
from langchain_core.messages import HumanMessage
from langchain_openai import ChatOpenAI
from langgraph.checkpoint.memory import MemorySaver
from langgraph.prebuilt import create_react_agent

load_dotenv()

DEFAULT_KREWE_PREDICT_URL = "https://krewe-orchestrator-production.up.railway.app/v2/predict"
DEFAULT_MAX_PAYMENT_USDC = 0.10


def initialize_agent(config: CdpEvmWalletProviderConfig):
    """Initialize the LangChain agent with AgentKit + krewe x402 inference.

    Args:
        config: Configuration for the CDP EVM Server Wallet Provider.

    Returns:
        tuple[Agent, CdpEvmWalletProvider]: The initialized agent and wallet provider.
    """
    llm = ChatOpenAI(model="gpt-4o-mini")

    wallet_provider = CdpEvmWalletProvider(
        CdpEvmWalletProviderConfig(
            api_key_id=config.api_key_id,
            api_key_secret=config.api_key_secret,
            wallet_secret=config.wallet_secret,
            network_id=config.network_id,
            address=config.address,
            idempotency_key=config.idempotency_key,
            rpc_url=config.rpc_url,
        )
    )

    krewe_url = os.getenv("KREWE_PREDICT_URL", DEFAULT_KREWE_PREDICT_URL)
    max_payment_usdc = float(os.getenv("KREWE_MAX_PAYMENT_USDC", DEFAULT_MAX_PAYMENT_USDC))

    agentkit = AgentKit(
        AgentKitConfig(
            wallet_provider=wallet_provider,
            action_providers=[
                cdp_api_action_provider(),
                cdp_evm_wallet_action_provider(),
                erc20_action_provider(),
                wallet_action_provider(),
                x402_action_provider(
                    X402Config(
                        registered_services=[krewe_url],
                        allow_dynamic_service_registration=False,
                        max_payment_usdc=max_payment_usdc,
                        registered_facilitators={},
                    )
                ),
            ],
        )
    )

    tools = get_langchain_tools(agentkit)

    memory = MemorySaver()
    agent_config = {"configurable": {"thread_id": "krewe Inference Chatbot Example"}}

    return (
        create_react_agent(
            llm,
            tools=tools,
            checkpointer=memory,
            state_modifier=(
                f"You are an onchain agent with access to:\n"
                f"  - A CDP EVM wallet on Base.\n"
                f"  - The krewe inference network at {krewe_url}, called via the x402 "
                f"action provider. krewe is a decentralized AI inference network — your "
                f"wallet pays USDC per call (cap: ${max_payment_usdc}). The supported "
                f"job kinds are text.structure (regex/JSON extraction, $0.005), "
                f"text.embed (sentence embeddings, $0.01), web.scrape (clean HTML "
                f"fetch, $0.02), and text.complete (small-LM completion, $0.05).\n\n"
                f"When a user asks for structured extraction, embeddings, a clean "
                f"scrape, or a quick small-model completion, prefer routing through "
                f"krewe via the x402 service — it pays krewe miners and gets a 2-of-3 "
                f"consensus output.\n\n"
                f'Always send the body: {{"kind": "...", "payload": <kind-specific JSON>}}. '
                f"If a call returns 402 'payment-required', the x402 provider handles "
                f"signing and on-chain settlement automatically — just retry once via "
                f"the same tool.\n\n"
                f"Be concise. If you can't do something with your current tools, say so."
            ),
        ),
        wallet_provider,
    ), agent_config


def setup():
    """Set up the agent with persistent wallet storage.

    Returns:
        tuple[Agent, dict]: The initialized agent and its configuration.
    """
    network_id = os.getenv("NETWORK_ID", "base-mainnet")
    wallet_file = f"wallet_data_{network_id.replace('-', '_')}.txt"

    # Load existing wallet data if available
    wallet_data: dict = {}
    if os.path.exists(wallet_file):
        try:
            import json

            with open(wallet_file) as f:
                wallet_data = json.load(f)
                print(f"Loading existing wallet from {wallet_file}")
        except json.JSONDecodeError:
            print(f"Warning: Invalid wallet data for {network_id}")
            wallet_data = {}

    wallet_address = wallet_data.get("address") or os.getenv("ADDRESS") or None

    config = CdpEvmWalletProviderConfig(
        api_key_id=os.getenv("CDP_API_KEY_ID"),
        api_key_secret=os.getenv("CDP_API_KEY_SECRET"),
        wallet_secret=os.getenv("CDP_WALLET_SECRET"),
        network_id=network_id,
        address=wallet_address,
        idempotency_key=(os.getenv("IDEMPOTENCY_KEY") if not wallet_address else None),
        rpc_url=os.getenv("RPC_URL"),
    )

    (agent_executor, wallet_provider), agent_config = initialize_agent(config)

    import json

    new_wallet_data = {
        "address": wallet_provider.get_address(),
        "network_id": network_id,
        "created_at": time.strftime("%Y-%m-%d %H:%M:%S")
        if not wallet_data
        else wallet_data.get("created_at"),
    }
    with open(wallet_file, "w") as f:
        json.dump(new_wallet_data, f, indent=2)
        print(f"Wallet data saved to {wallet_file}")

    return (agent_executor, agent_config)


def run_autonomous_mode(agent_executor, config, interval=30):
    """Run the agent autonomously — picks a creative krewe inference job each tick."""
    print("Starting autonomous mode...")
    while True:
        try:
            thought = (
                "Pick an interesting short snippet of text or a public URL and route a "
                "krewe job through the x402 service to do something useful with it "
                "(extract structured data, embed, summarize, or scrape). Tell me the "
                "kind you chose, the payment cost, and the resulting output."
            )
            for chunk in agent_executor.stream(
                {"messages": [HumanMessage(content=thought)]}, config
            ):
                if "agent" in chunk:
                    print(chunk["agent"]["messages"][0].content)
                elif "tools" in chunk:
                    print(chunk["tools"]["messages"][0].content)
                print("-------------------")
            time.sleep(interval)
        except KeyboardInterrupt:
            print("Goodbye Agent!")
            sys.exit(0)


def run_chat_mode(agent_executor, config):
    """Run the agent interactively based on user input."""
    print("Starting chat mode... Type 'exit' to end.")
    print(
        'Try: "Extract emails and dates from this text: '
        'Email hi@krewe.world by 2026-05-17"'
    )
    while True:
        try:
            user_input = input("\nPrompt: ")
            if user_input.lower() == "exit":
                break
            for chunk in agent_executor.stream(
                {"messages": [HumanMessage(content=user_input)]}, config
            ):
                if "agent" in chunk:
                    print(chunk["agent"]["messages"][0].content)
                elif "tools" in chunk:
                    print(chunk["tools"]["messages"][0].content)
                print("-------------------")
        except KeyboardInterrupt:
            print("Goodbye Agent!")
            sys.exit(0)


def choose_mode():
    """Choose whether to run in autonomous or chat mode based on user input."""
    while True:
        print("\nAvailable modes:")
        print("1. chat    - Interactive chat mode")
        print("2. auto    - Autonomous action mode")
        choice = input("\nChoose a mode (enter number or name): ").lower().strip()
        if choice in ["1", "chat"]:
            return "chat"
        elif choice in ["2", "auto"]:
            return "auto"
        print("Invalid choice. Please try again.")


def main():
    """Start the chatbot agent."""
    load_dotenv()
    agent_executor, agent_config = setup()
    mode = choose_mode()
    if mode == "chat":
        run_chat_mode(agent_executor=agent_executor, config=agent_config)
    elif mode == "auto":
        run_autonomous_mode(agent_executor=agent_executor, config=agent_config)


if __name__ == "__main__":
    print("Starting krewe-inference Agent...")
    main()

File: typescript/examples/langchain-krewe-inference-chatbot/chatbot.ts
import {
  AgentKit,
  CdpEvmWalletProvider,
  walletActionProvider,
  cdpApiActionProvider,
  cdpEvmWalletActionProvider,
  erc20ActionProvider,
  x402ActionProvider,
} from "@coinbase/agentkit";
import { getLangChainTools } from "@coinbase/agentkit-langchain";
import { HumanMessage } from "@langchain/core/messages";
import { MemorySaver } from "@langchain/langgraph";
import { createAgent } from "langchain";
import { ChatOpenAI } from "@langchain/openai";
import * as dotenv from "dotenv";
import * as readline from "readline";

dotenv.config();

const DEFAULT_KREWE_PREDICT_URL = "https://krewe-orchestrator-production.up.railway.app/v2/predict";
const DEFAULT_MAX_PAYMENT_USDC = 0.1;

/**
 * Validates that required environment variables are set.
 *
 * @throws {Error} - If required environment variables are missing
 * @returns {void}
 */
function validateEnvironment(): void {
  const requiredVars = [
    "OPENAI_API_KEY",
    "CDP_API_KEY_ID",
    "CDP_API_KEY_SECRET",
    "CDP_WALLET_SECRET",
  ];
  const missing = requiredVars.filter(name => !process.env[name]);
  if (missing.length > 0) {
    console.error("Error: Required environment variables are not set");
    missing.forEach(name => console.error(`${name}=your_${name.toLowerCase()}_here`));
    process.exit(1);
  }
  if (!process.env.NETWORK_ID) {
    console.warn("Warning: NETWORK_ID not set, defaulting to base-mainnet (krewe is mainnet-only today)");
  }
}

validateEnvironment();

/**
 * Initialize the agent with AgentKit + the krewe x402 inference service.
 *
 * @returns Agent executor and config
 */
async function initializeAgent() {
  const llm = new ChatOpenAI({ model: "gpt-4o-mini" });

  const networkId = process.env.NETWORK_ID || "base-mainnet";

  const walletProvider = await CdpEvmWalletProvider.configureWithWallet({
    apiKeyId: process.env.CDP_API_KEY_ID,
    apiKeySecret: process.env.CDP_API_KEY_SECRET,
    walletSecret: process.env.CDP_WALLET_SECRET,
    idempotencyKey: process.env.IDEMPOTENCY_KEY,
    address: process.env.ADDRESS as `0x${string}` | undefined,
    networkId,
    rpcUrl: process.env.RPC_URL,
  });

  // krewe is an x402-paywalled inference network on Base mainnet. The
  // built-in x402ActionProvider does the entire handshake: 402 challenge,
  // EIP-3009 signature, on-chain USDC settlement, and the actual call.
  //
  // Pricing per call (USDC base units, 6 decimals):
  //   $0.005 text.structure   ·   $0.01  text.embed
  //   $0.02  web.scrape       ·   $0.05  text.complete
  //
  // Every USDC paid here is swept hourly by krewe's buyback service, swapped
  // for $KREW on Uniswap V4, and deposited back into the registry's reward
  // pool — so each call directly funds the miners that answered it.
  const kreweUrl = process.env.KREWE_PREDICT_URL ?? DEFAULT_KREWE_PREDICT_URL;
  const maxPaymentUsdc = Number(process.env.KREWE_MAX_PAYMENT_USDC ?? DEFAULT_MAX_PAYMENT_USDC);

  const actionProviders = [
    walletActionProvider(),
    cdpApiActionProvider(),
    cdpEvmWalletActionProvider(),
    erc20ActionProvider(),
    x402ActionProvider({
      registeredServices: [kreweUrl],
      allowDynamicServiceRegistration: false,
      maxPaymentUsdc,
      registeredFacilitators: {},
    }),
  ];

  const agentkit = await AgentKit.from({ walletProvider, actionProviders });
  const tools = await getLangChainTools(agentkit);

  const memory = new MemorySaver();
  const agentConfig = { configurable: { thread_id: "krewe Inference Chatbot Example" } };

  const agent = createAgent({
    model: llm,
    tools,
    checkpointer: memory,
    systemPrompt: `
You are an onchain agent with access to:
  • A CDP EVM wallet on Base.
  • The krewe inference network at ${kreweUrl}, accessed via the x402
    action provider. krewe is a decentralized AI inference network — your
    wallet pays USDC per call (cap: $${maxPaymentUsdc}). The supported job
    kinds are text.structure (regex/JSON extraction, $0.005), text.embed
    (sentence embeddings, $0.01), web.scrape (clean HTML fetch, $0.02),
    and text.complete (small-LM completion, $0.05).

When a user asks for structured extraction, embeddings, a clean scrape,
or a quick small-model completion, prefer routing through krewe via the
x402 service — it pays krewe miners and gets a 2-of-3 consensus output.

Always include the relevant body when calling the krewe endpoint:
  { "kind": "text.structure" | "text.embed" | "web.scrape" | "text.complete",
    "payload": <kind-specific JSON> }

If a call returns 402 'payment-required', the x402 provider handles the
signature + settlement automatically — just retry once via the same tool.

Be concise. If you can't do something with your current tools, say so.
`,
  });

  return { agent, config: agentConfig };
}

/**
 * Run the agent interactively based on user input.
 *
 * @param agent - The agent executor
 * @param config - Agent configuration
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function runChatMode(agent: any, config: any) {
  console.log("Starting chat mode... Type 'exit' to end.");
  console.log("Try: \"Extract emails and dates from this text: Email hi@krewe.world by 2026-05-17\"");

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const question = (prompt: string): Promise<string> =>
    new Promise(resolve => rl.question(prompt, resolve));

  try {
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const userInput = await question("\nPrompt: ");
      console.log("-------------------");
      if (userInput.toLowerCase() === "exit") break;

      const stream = await agent.stream({ messages: [new HumanMessage(userInput)] }, config);

      for await (const chunk of stream) {
        if ("model_request" in chunk) {
          const response = chunk.model_request.messages[0].content;
          if (response !== "") console.log("\n Response: " + response);
        }
        if ("tools" in chunk) {
          for (const tool of chunk.tools.messages) {
            console.log("Tool " + tool.name + ": " + tool.content);
          }
        }
      }
      console.log("-------------------");
    }
  } catch (error) {
    if (error instanceof Error) console.error("Error:", error.message);
    process.exit(1);
  } finally {
    rl.close();
  }
}

/**
 * Run the agent autonomously — picks a creative krewe inference job each
 * tick to exercise the network.
 *
 * @param agent - The agent executor
 * @param config - Agent configuration
 * @param interval - Time interval between actions in seconds
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function runAutonomousMode(agent: any, config: any, interval = 30) {
  console.log("Starting autonomous mode...");

  // eslint-disable-next-line no-constant-condition
  while (true) {
    try {
      const thought =
        "Pick an interesting short snippet of text or a public URL and route a krewe job through " +
        "the x402 service to do something useful with it (extract structured data, embed, summarize, " +
        "or scrape). Tell me the kind you chose, the payment cost, and the resulting output.";

      const stream = await agent.stream({ messages: [new HumanMessage(thought)] }, config);

      for await (const chunk of stream) {
        if ("model_request" in chunk) {
          const response = chunk.model_request.messages[0].content;
          if (response !== "") console.log("\n Response: " + response);
        }
        if ("tools" in chunk) {
          for (const tool of chunk.tools.messages) {
            console.log("Tool " + tool.name + ": " + tool.content);
          }
        }
      }
      console.log("-------------------");

      await new Promise(resolve => setTimeout(resolve, interval * 1000));
    } catch (error) {
      if (error instanceof Error) console.error("Error:", error.message);
      process.exit(1);
    }
  }
}

/**
 * Choose whether to run in autonomous or chat mode based on user input.
 *
 * @returns Selected mode
 */
async function chooseMode(): Promise<"chat" | "auto"> {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const question = (prompt: string): Promise<string> =>
    new Promise(resolve => rl.question(prompt, resolve));

  // eslint-disable-next-line no-constant-condition
  while (true) {
    console.log("\nAvailable modes:");
    console.log("1. chat    - Interactive chat mode");
    console.log("2. auto    - Autonomous action mode");

    const choice = (await question("\nChoose a mode (enter number or name): "))
      .toLowerCase()
      .trim();
    if (choice === "1" || choice === "chat") {
      rl.close();
      return "chat";
    } else if (choice === "2" || choice === "auto") {
      rl.close();
      return "auto";
    }
    console.log("Invalid choice. Please try again.");
  }
}

/**
 * Start the chatbot agent.
 */
async function main() {
  try {
    const { agent, config } = await initializeAgent();
    const mode = await chooseMode();
    if (mode === "chat") await runChatMode(agent, config);
    else await runAutonomousMode(agent, config);
  } catch (error) {
    if (error instanceof Error) console.error("Error:", error.message);
    process.exit(1);
  }
}

if (require.main === module) {
  console.log("Starting krewe-inference agent...");
  main().catch(error => {
    console.error("Fatal error:", error);
    process.exit(1);
  });
}
